File: finalyearproject/repos/dkd-b2c-order/src/persistence/persistence.ts
// DB abstraction + tx helper + repository base + migrations. The concrete driver (pg) is wired at
// the integration point. No business repositories.
export interface Tx { exec(sql: string, args?: unknown[]): Promise<void>; }
export interface Db {
  ping(): Promise<void>;
  withTx<T>(fn: (tx: Tx) => Promise<T>): Promise<T>;
  close(): Promise<void>;
}
export interface Migrator { apply(): Promise<void>; }

export abstract class RepositoryBase {
  constructor(protected readonly db: Db) {}
  protected inTx<T>(fn: (tx: Tx) => Promise<T>): Promise<T> { return this.db.withTx(fn); }
}

File: finalyearproject/repos/dkd-logistics/_stripped/persistence/persistence.go
package persistence

import "context"

// DB is the persistence abstraction. The concrete driver (pgx for Postgres) is wired at the
// integration point. No business repositories are defined here.
type DB interface {
	Ping(ctx context.Context) error
	WithTx(ctx context.Context, fn func(tx Tx) error) error
	Close() error
}

type Tx interface {
	Exec(ctx context.Context, sql string, args ...any) error
}

// Repository is the base every context repository embeds; it carries the DB handle and the tx helper.
type Repository struct {
	DB DB
}

func (r Repository) InTx(ctx context.Context, fn func(tx Tx) error) error {
	return r.DB.WithTx(ctx, fn)
}

// Migrator runs ordered, idempotent schema migrations at startup. The driver supplies Apply.
type Migrator interface {
	Apply(ctx context.Context) error
}
